File: situational-awareness/backend/app/schemas/discovery.py
from datetime import datetime

from pydantic import BaseModel, IPvAnyNetwork

from app.schemas.campus import ScannerZoneRead
from app.schemas.common import PageMeta
from app.db.models.enums import DiscoveryJobStatus
from app.schemas.common import ORMModel


class DiscoveryJobCreate(BaseModel):
    cidr: IPvAnyNetwork
    label: str | None = None
    runner_asset_id: str | None = None
    scanner_zone_id: str | None = None


class DiscoveryJobRead(ORMModel):
    id: str
    cidr: IPvAnyNetwork
    status: DiscoveryJobStatus
    label: str | None
    scanner_zone_id: str | None = None
    started_at: datetime | None
    finished_at: datetime | None
    created_at: datetime
    summary_json: dict


class DiscoveryJobCreateResponse(BaseModel):
    job: DiscoveryJobRead
    task_id: str
    status: str
    reused: bool = False


class DiscoveryJobListResponse(BaseModel):
    items: list[DiscoveryJobRead]
    meta: PageMeta


class DiscoveryRunnerOptionRead(BaseModel):
    runner_id: str | None = None
    asset_id: str
    asset_ip: str | None = None
    asset_hostname: str | None = None
    status: str
    install_status: str
    version: str | None = None
    scanner_zone_id: str | None = None
    last_seen_at: str | None = None
    detected_os: str | None = None
    detected_arch: str | None = None
    compatibility_issues: list[str] = []
    capabilities_json: dict = {}


class DiscoverySchedulingOptionRead(BaseModel):
    recommended_zone_ids: list[str] = []
    scanner_zones: list[ScannerZoneRead] = []
    runner_assets: list[DiscoveryRunnerOptionRead] = []

File: situational-awareness/frontend/src/types/discovery.ts
export type DiscoveryJob = {
  id: string;
  cidr: string;
  status: "pending" | "running" | "completed" | "failed";
  label: string | null;
  scanner_zone_id?: string | null;
  started_at: string | null;
  finished_at: string | null;
  created_at: string;
  summary_json: Record<string, unknown>;
};

export type DiscoveryJobCreateResponse = {
  job: DiscoveryJob;
  task_id: string;
  status: "pending" | "reused";
  reused: boolean;
};

export type DiscoveryJobListResponse = {
  items: DiscoveryJob[];
  meta: {
    total: number;
    page: number;
    page_size: number;
  };
};

export type DiscoverySchedulingOption = {
  recommended_zone_ids: string[];
  scanner_zones: Array<{
    id: string;
    name: string;
    zone_type: "office" | "dormitory" | "wireless" | "server" | "iot" | "custom";
    description: string | null;
    priority: number;
    enabled: boolean;
    cidrs_json: string[];
    default_scan_profile_json: Record<string, unknown>;
    allowed_data_source_types_json: string[];
    created_at: string;
    updated_at: string;
  }>;
  runner_assets: Array<{
    runner_id: string | null;
    asset_id: string;
    asset_ip: string | null;
    asset_hostname: string | null;
    status: string;
    install_status: string;
    version: string | null;
    scanner_zone_id: string | null;
    last_seen_at: string | null;
    detected_os: string | null;
    detected_arch: string | null;
    compatibility_issues: string[];
    capabilities_json: Record<string, unknown>;
  }>;
};
